vignettes/ipums-read.Rmd

```diff
@@ -0,0 +1,390 @@
+those in the DDI:
+
+```{r}
+cps_data2 <- set_ipums_var_attributes(cps_data2, cps_ddi)
+
+ipums_val_labels(cps_data2$MONTH)
+```
+
+### Hierarchical Extracts
+
+IPUMS microdata can come in either "rectangular" or "hierarchical"
+format.
+
+Rectangular data are transformed such that each row of data represents
+only one type of record. For instance, each row will represent a person
+record, and all household-level information for that person will be
+included in the same row. (This is the case for the CPS example above.)
+
+Hierarchical data have records of different types interspersed in a
+single file. For instance, a household record will be included in its
+own row followed by the person records associated with that household.
+
+Hierarchical data can be loaded in list format or long format.
+`read_ipums_micro()` will read in long format:
+
+```{r}
+cps_hier_ddi <- read_ipums_ddi(ipums_example("cps_00010.xml"))
+
+read_ipums_micro(cps_hier_ddi)
+```
+
+This structure contains a single row for each record, regardless of
+record type. Variables that do not apply to a particular record type
+will be filled with `NA` in rows of that record type.
+
+To read data in list format, use `read_ipums_micro_list()`. This splits
+the record types into a list where each element contains all the records
+for a given type:
+
+```{r}
+read_ipums_micro_list(cps_hier_ddi)
+```
+
+`read_ipums_micro()` and `read_ipums_micro_list()` also support partial
+loading by selecting only a subset of columns or a limited number of
+rows. See the documentation for more details about other options.
+
+## Reading Aggregate Data Extracts
+
+Unlike microdata projects, NHGIS extracts provide their data and
+metadata files bundled into a single .zip archive. `read_nhgis()`
+anticipates this structure and can read data files directly from this
+file without the need to manually extract the files:
+
+```{r}
+nhgis_ex1 <- ipums_example("nhgis0972_csv.zip")
+
+nhgis_data <- read_nhgis(nhgis_ex1)
+
+nhgis_data
+```
+
+Like microdata extracts, the data include variable-level metadata, where
+available:
+
+```{r}
+attributes(nhgis_data$D6Z001)
+```
+
+Variable metadata for NHGIS data are slightly different than those
+provided by microdata products. First, they come from a .txt codebook
+file rather than an .xml DDI file. Codebooks can still be loaded into an
+`ipums_ddi` object, but many fields that do not apply to aggregate data
+will be empty. In general, NHGIS codebooks provide only variable labels
+and descriptions along with citation information.
+
+```{r}
+nhgis_cb <- read_nhgis_codebook(nhgis_ex1)
+
+# Most useful metadata for NHGIS is for variable labels:
+ipums_var_info(nhgis_cb) %>%
+  select(var_name, var_label, var_desc)
+```
+
+By design, NHGIS codebooks are human-readable. To view the codebook
+contents themselves without converting to an `ipums_ddi` object, set
+`raw = TRUE`.
+
+```{r}
+read_nhgis_codebook(nhgis_ex1, raw = TRUE)[1:20]
```

Comment:
Would it be useful to provide an option to open the codebook file with `file.edit()`?

Reply:
This is a bit sticky because of the fact that we're working inside zip files. We could handle those connections if we wanted, but this is pretty similar to using the current `raw = TRUE`, so for now I may put this on the backburner.